app/gui2/shared/ast/tree.ts

```diff
@@ -306,6 +326,99 @@ export abstract class MutableAst extends Ast {
+    const newValue = rewriteFieldRefs(value, f)
+    if (newValue !== undefined) {
+      ast.fields.set(key as any, newValue)
+      fieldsChanged += 1
+    }
+  }
+  return fieldsChanged
+}
+export function syncFields(ast1: MutableAst, ast2: Ast, f: (id: AstId) => AstId | undefined) {
+  const expressionMetadata =
+    ast1 instanceof MutableAssignment ? ast1.expression.fields.get('metadata').clone() : undefined
+  for (const [key, value] of (ast2.fields as any as Map<string, FieldData>).entries()) {
+    if (astFieldKeys.includes(key as any)) continue
+    const changedValue = rewriteFieldRefs(value, f)
+    const newValue = changedValue ?? value
+    if (!fieldEqual(ast1.fields.get(key as any), newValue)) ast1.fields.set(key as any, newValue)
+  }
+  if (ast1 instanceof MutableAssignment) ast1.expression.fields.set('metadata', expressionMetadata!)
+}
+
+/** @internal */
+export function rewriteFieldRefs(field: FieldData, f: (id: AstId) => AstId | undefined): FieldData {
+  if (field === undefined) return field
+  if ('node' in field) {
+    const child = field.node as AstId | SyncTokenId
+    if (isTokenId(child)) return
+    const newValue = f(child)
+    if (newValue !== undefined) {
+      field.node = newValue
+      return field
+    }
+  } else if (Array.isArray(field)) {
+    let fieldChanged = false
+    field.forEach((subfield, i) => {
+      const newValue = rewriteFieldRefs(subfield, f)
+      if (newValue !== undefined) {
+        field[i] = newValue
+        fieldChanged = true
+      }
+    })
+    if (fieldChanged) return field
+  } else {
+    const fieldObject = field as FieldObject
+    let fieldChanged = false
+    for (const [key, value] of Object.entries(fieldObject)) {
+      const newValue = rewriteFieldRefs(value, f)
+      if (newValue !== undefined) {
+        fieldObject[key] = newValue
+        fieldChanged = true
+      }
+    }
+    if (fieldChanged) return fieldObject
+  }
+}
+function fieldEqual(field1: FieldData, field2: FieldData): boolean {
+  if (field1 === undefined) return field2 === undefined
+  if (field2 === undefined) return false
+  if ('node' in field1 && 'node' in field2) {
+    return field1['whitespace'] === field2['whitespace'] && field1.node === field2.node
+  } else if ('node' in field1 || 'node' in field2) {
+    return false
+  } else if (Array.isArray(field1) && Array.isArray(field2)) {
+    return (
+      field1.length === field2.length && field1.every((value1, i) => fieldEqual(value1, field2[i]))
+    )
+  } else if (Array.isArray(field1) || Array.isArray(field2)) {
+    return false
+  } else {
+    const fieldObject1 = field1 as FieldObject
+    const fieldObject2 = field2 as FieldObject
+    const keys = new Set<string>()
+    for (const key of Object.keys(fieldObject1)) keys.add(key)
+    for (const key of Object.keys(fieldObject2)) keys.add(key)
+    for (const key of keys) if (fieldObject1[key] !== fieldObject2[key]) return false
+    return true
+  }
+}
```

Comment:
Those functions contain very optimistic typecasts. All this seems to be a consequence of non-ideal `FieldData` definition. I don't have an easy solution here, but I'd really prefer to remove as much of those forced typecasts as possible. In general, more proper type validation functions would be great.